### PR TITLE
ROX-23692: Profile clusters results table

### PR DESCRIPTION
**Pull request by bradr5**
## Description

Creates a table to display a single profiles check results

Notable changes:

- Unlike the checks results table, table state lives at the page level
- Taking advantage of the `getDistanceStrictAsPhrase` utility function to display the "Last scanned" column
- Cluster details route changed to use `clusterId` as unique identifier instead of `clusterName` to match api
- Sorting, pagination, filtering will come later

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
Screenshot 2024-05-13 at 3 31 59 PM

### CODE REVIEW

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx [hunk 1]
```diff
@@ -1,44 +1,56 @@
...
+import CoveragesToggleGroup from './CoveragesToggleGroup';
+import ProfileClustersTable from './ProfileClustersTable';
 
 function ProfileClustersPage() {
-    const history = useHistory();
     const { profileName } = useParams();
+    const [currentDatetime, setCurrentDatetime] = useState<Date>(new Date());
+
+    const fetchProfileClusters = useCallback(
+        () => getComplianceClusterStats(profileName),
```
Comment (pedrottimark) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:19]: Here is an observation, not necessarily a recommendation.
* On one side of the coin `useState<Date>(new Date())` might bend the React rule of no side effects, because it has a different value on every call, unlike `''` or `false` or `null`.
* On other side of the coin `setCurrentDatetime(new Date());` is explicitly in `useEffect` hook.

An alternative might call a wrapper function for `getComplianceClusterStats` that can return `{ currentDatetime: Date, profileClusters: ListComplianceClusterOverallStatsResponse }` via `new Date()` in a `then` block.

Comment (sachaudh) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:19]: I don't think that `useState<Date>(new Date())` will necessarily cause side effects. But the `new Date()` expression will be evaluated on every render, which would produce a new `Date` object. That might be unnecessary. There is a trick that will allow us to only evaluate it once though. You can do the following:

```
const [currentDatetime, setCurrentDatetime] = useState(() => new Date());
```

Also, I don't think you'll need to add the `<Date>` since Typescript will be able to determine the type based on the initial state.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx [hunk 2]
```diff
@@ -1,44 +1,56 @@
...
-                    Go to single cluster (ClusterDetailsPage)
-                </Button>
+            <PageSection variant="default">
+                <PageSection variant="light" component="div">
+                    <Title headingLevel="h2">Profile results</Title>
+                    <Divider />
+                    <ProfileClustersTable
+                        currentDatetime={currentDatetime}
+                        profileName={profileName}
+                        tableState={tableState}
```
Comment (sachaudh) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:51]: What do you think about keeping the content of `ProfileClustersTable` within this component? Sometimes I've found that creating too many layers to components in the beginning can make it hard to follow and debug. You need to jump around a bit to find what you need. For this case, I think it would make sense to keep things centralized for now. 

Example 1:
```
<Page>
    <PageHeader />
    <Table />
</Page>
```



Example 2:
```
<Page>
    <PageHeader />
    <Table>
        <TableHead>
            <Tr>
                <Th>...</Th>
                <Th>...</Th>
                <Th>...</Th>
            </Tr>
        </TableHead>
        <TableBody>
           <Tr>
                <Td>...</Td>
                <Td>...</Td>
                <Td>...</Td>
            </Tr>
        </TableBody>
    </Table>
</Page>
```

For the first case, you'd have to dig into that component to see what's going on. In the second case, you can see everything at a first glance.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersTable.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                                const { passCount, failCount, otherCount, totalCount } =
+                                    getStatusCounts(checkStats);
+                                const passPercentage = calculateCompliancePercentage(
+                                    passCount,
+                                    totalCount
+                                );
+                                const progressBarId = `progress-bar-${clusterId}`;
+                                const firstDiscoveredAsPhrase = getDistanceStrictAsPhrase(
+                                    lastScanTime,
+                                    currentDatetime
```
Comment (sachaudh) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersTable.tsx:90]: @bradr5 If we did `new Date()` here instead of keeping track of `currentDatetime` wouldn't that equate to the same result we want? Ideally, this `date.map` won't re-evaluate unless the data refreshes, right?

**Review by sachaudh — APPROVED** [review]

**Review by pedrottimark — APPROVED** [review]

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx (hunk 2 above)
Comment (bradr5) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:51]: It's a good thought and I agree that creating too many layers can make things complex but I think I'd argue the opposite here. This feels like a clear separation of concerns and a designated "Table" component would hopefully increase readability. Otherwise, all table-specific information would clutter up the "Page" component.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx (hunk 1 above)
Comment (bradr5) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:19]: I like Mark's suggestion because the date would be tightly couple with `profileClusters` data being fetched. We ran into a typescript issue when trying to implement it though. _We_ can confidently say date would never be displayed if data isn't fetched but typescript doesn't know that. 

@sachaudh can you think of an instance where date would be re-evaluated without `profileClusters` being fetched? That would be an issue but I tried to think through edge cases and don't think that should happen.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersTable.tsx (hunk 3 above)
Comment (bradr5) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersTable.tsx:90]: That's an option, I would just worry about us adding state to the `ProfileClustersTable` component that would cause Date to be re-evaluated without data being re-fetched. I'm wanting to tightly couple `currentDatetime` with `getComplianceClusterStats` which is why I like Mark's suggestion.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx (hunk 2 above)
Comment (sachaudh) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/Coverage/ProfileClustersPage.tsx:51]: Hmm, can you explain why you think it separates concerns? It seems to only render the table without any state logic. Because of that, I believe it would be more readable in the parent component, as this keeps the logic inline and avoids context switching to understand the component's purpose. It feels like an unnecessary abstraction right now. But I'm sure you have a bigger picture in mind for the component, so I'm good with keeping it like this. Just wanted to raise my concerns.
